external/fv3fit/fv3fit/pytorch/graph.py

```diff
@@ -0,0 +1,183 @@
+        return set(self.input_variables).union(self.output_variables)
+
+
+def get_normalizer(sample: Mapping[str, np.ndarray]):
+    scalers = {}
+    for name, array in sample.items():
+        s = StandardScaler()
+        s.fit(array)
+        scalers[name] = s
+
+    def scale(sample: Mapping[str, np.ndarray]):
+        output = {**sample}
+        for name, array in sample.items():
+            output[name] = scalers[name].normalize(array)
+        return output
+
+    return scale
+
+
+# TODO: Still have to handle forcing
+
+
+@register_training_function("graph", GraphHyperparameters)
+def train_graph_model(
+    hyperparameters: GraphHyperparameters,
+    train_batches: tf.data.Dataset,
+    validation_batches: Optional[tf.data.Dataset],
+):
+
+    """
+    Train a graph network.
+
+    Args:
+        train_batches: training data, as a dataset of Mapping[str, tf.Tensor]
+        validation_batches: validation data, as a dataset of Mapping[str, tf.Tensor]
+        build_model: the function which produces the pytorch model
+            from input and output samples. The models returned must take a list of
+            tensors as input and return a list of tensors as output.
+        input_variables: names of inputs for the pytorch model
+        output_variables: names of outputs for the pytorch model
+        n_epoch: number of epochs
+    """
+
+    # use transforms to get correct variables, correct dimensions
+    processed_train_batches = train_batches.map(apply_to_mapping(ensure_nd(3)))
+
+    sample = next(
+        iter(
+            processed_train_batches.unbatch().batch(
+                hyperparameters.normalization_fit_samples
+            )
+        )
+    )
+
+    normalizer = get_normalizer(sample)
+
+    get_Xy = curry(get_Xy_dataset)(
+        input_variables=hyperparameters.input_variables,
+        output_variables=hyperparameters.output_variables,
+        n_dims=3,
+    )
+
+    if validation_batches is not None:
+        validation_batches = validation_batches.map(apply_to_mapping(ensure_nd(3)))
+        validation_batches = validation_batches.map(normalizer)
+        val_Xy = get_Xy(data=validation_batches)
+    else:
+        val_Xy = None
+
+    processed_train_batches = processed_train_batches.map(normalizer)
+    train_Xy = get_Xy(data=processed_train_batches)
+
+    train_model = build_model(hyperparameters)
+    optimizer = hyperparameters.optimizer_config
+
+    hyperparameters.training_loop.fit_loop(
+        hyperparameters.loss,
+        train_model=train_model,
+        train_data=train_Xy,
+        validation_data=val_Xy,
+        optimizer=optimizer.instance(train_model.parameters()),
+    )
+
+    predictor = PytorchModel(
+        input_variables=hyperparameters.input_variables,
+        output_variables=hyperparameters.output_variables,
+        model=train_model,
+        unstacked_dims=("z",),
+    )
+    return predictor
+
+
+def build_model(config: GraphHyperparameters):
```

Comment:
Please refactor this method to take in only `config.build_graph` and `config.graph_network`, it doesn't need access to the other configuration and this change would make that clear.